enhancements/pipeline-tooling.md

```diff
@@ -0,0 +1,267 @@
+# Pipeline Authors have Tools to build, modify, and consume the latest bundle and registry formats
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Summary
+
+There are several internal pipelines that deal with OLM operator data, all of which need to build or consume the newer bundle and registry formats. We can build new bundle and registry formats as per the other 4.4 epics, but if no one internally is using them we will not derive much of the desired value from them.
+
+### Goals
+
+* Maintainers of Operator Registries can leverage container registries so that they can host Operator bundles and indexes of all available bundles in this catalog in the form of container images
+* Maintainers of Operator Registries can use a tool to:
+    * create a bundle index so that it is easy to employ in a CI environment.
+    * migrate to and from both old and new format.
+    * merge indexes
+    * delete the latest version of an operator from an index
+    * download the contents of index images
+    * specify related images for CVE updates.
+    * validate the contents of an index before publishing it.
+    * package/release/version tool(s)
+* Pipeline admins can publish multiple applications simultaneously without conflicts.
+* Developers or QE can use downstream images in a dev/stage/qe repository
+* There is sufficient documentation for operator owners and maintainers of operatory registries.
+* Design the migration of Community Operators to the new bundle format.
+
+### Non-Goals
+
+* Migrate Community Operators / operatorhub.io to the new bundle format.
+* Pipeline admins rely on the tooling to extrapolate defined selective metadata from full set of each application operator manifests and include that in a combined catalog for index object
+* Pipeline admins can update official catalog object with newly created single application operator and publish
+
+---
+
+## User Stories
+
+### Migrate to and from each registry format
+
+As an Operator Registry maintainer, I want to continue to push updates to old (appregistry-based) catalogs after having migrated to the new bundle format without having to maintain two repositories.
+
+#### Design Proposal
+
+Add a command to `opm index` and `opm alpha bundle` called `migrate` which:
```

Comment:
This seems to imply that the new pipeline will switch to outputting bundles and then we'll convert those back to the format for appregistry.

What about adding a "fork" in the pipeline?

```
           - appregistry
         /
git --
         \ 
           - bundle image
```
is this too simplistic a view to be useful?
?

Reply:
Hmm, the appregistry side of the "fork" contains all versions, while the bundle side only contains one. So long as we have to support appregistry, do we just need to keep old manifests versions in the repo (and skips).

Reply:
I think you're hitting on exactly what we need to nail down. This will need to be a "phased" migration of sorts, so that we can continuously deploy to appregistry until we are ready to flip the switch.

What are the phases of the migration (there may be different options here depending on the changes we make to the pipeline)?

I would image it would be something like:

 1. Set up a registry for storing bundle images
 1. Build appregistry and image bundles from the existing git repos. To start this will mean building bundles from the "old" filestructure, and deciding which ones need to be updated in the registry.
 1. Backfill the index images with data from the appregistry data (may not be necessary depending on how we do the previous step)
 1. Start building index images that contain the new bundles
 1. Flip openshift over to pull index images instead of appregistry (OperatorSource -> CatalogSource migration)
 1. When all openshift versions that use OperatorSources fall out of support, remove the appregistry side of the pipeline

Another option could be something like:

 1. Set up a registry for storing bundle images
 1. Introduce a new convention for storing manifests for operator repos, e.g. operator authors put manifests in `/manifests` in their git repo, but still keep their `/olm/1.0.0` and `/olm/1.0.1` folder.
 1. The pipeline requires that whatever is in `/manifests` also be correctly stored somewhere in the appregistry folder in their gitrepo
 1. We start buliding bundle images from the data in the new location, `/manifests`
 1. Backfill our bundle images from the appregistry data
 1. Start building index images from the bundle images
 1. Flip openshift over to pull index images instead of appregistry (OperatorSource -> CatalogSource migration)
 1. When all openshift versions that use OperatorSources fall out of support, remove the appregistry side of the pipeline

I'm sure there are more options here, and I'm assuming a lot about the pipelines, but this is a sort of strawman of what I was thinking.

I'm not sure how we continue to ship continuously to OperatorSource clusters if we don't do something along these lines.